### PR TITLE
`SqliteDosStorage`: Implement the storage backup

### PR DESCRIPTION
**Pull request by sphuber**

### CODE REVIEW

File: src/aiida/orm/implementation/storage_backend.py [hunk 1]
```diff
@@ -405,9 +405,18 @@ def backup(
...
         """
+        from aiida.common.exceptions import LockedProfileError, StorageBackupError
         from aiida.manage.configuration.settings import DEFAULT_CONFIG_FILE_NAME
+        from aiida.manage.profile_access import ProfileAccessManager
         from aiida.storage.log import STORAGE_LOGGER
 
+        # check that the AiiDA profile is not locked and request access for the duration of this backup process
+        # (locked means that possibly a maintenance operation is running that could interfere with the backup)
+        try:
+            ProfileAccessManager(self._profile).request_access()
```
Comment (danielhollas) [inline src/aiida/orm/implementation/storage_backend.py:416]: I am unfamiliar with profile locking -- does `request_access` lock the profile? If yes, where is the lock released? If no, is the locking performed later?

File: tests/storage/sqlite_dos/test_backend.py [hunk 2]
```diff
@@ -25,3 +25,15 @@ def test_archive_import(aiida_config, aiida_profile_factory):
         assert QueryBuilder().append(Node).count() == 0
         import_archive(get_archive_file('calcjob/arithmetic.add.aiida'))
         assert QueryBuilder().append(Node).count() > 0
+
+
+def test_backup(aiida_config, aiida_profile_factory, tmp_path, manager):
+    """Test the backup implementation."""
+    with aiida_profile_factory(aiida_config, storage_backend='core.sqlite_dos'):
+        storage = manager.get_profile_storage()
+        storage.backup(str(tmp_path))
```
Comment (danielhollas) [inline tests/storage/sqlite_dos/test_backend.py:34]: No strictly related to this PR but why doesn't storage backup accept Path?

File: tests/storage/sqlite_dos/test_backend.py [hunk 3]
```diff
@@ -25,3 +25,15 @@ def test_archive_import(aiida_config, aiida_profile_factory):
...
+        assert (tmp_path / 'config.json').exists()
+        assert (tmp_path / 'last-backup').is_symlink()
+        dirpath_backup = next(iter(tmp_path.glob('backup_*')))
```
Comment (danielhollas) [inline tests/storage/sqlite_dos/test_backend.py:37]: Maybe it would be better (both from testing and clarity perspective) to get this directory by resolving the `last-backup` symlink?

File: src/aiida/storage/sqlite_dos/backend.py [hunk 4]
```diff
@@ -225,3 +234,36 @@ def _get_mapper_from_entity(entity_type: 'EntityTypes', with_pk: bool):
...
+        dest: str,
+        keep: Optional[int] = None,
+    ):
+        try:
+            backup_manager = backup_utils.BackupManager(dest, keep=keep)
+            backup_manager.backup_auto_folders(lambda path, prev: self._backup_storage(backup_manager, path, prev))
+        except backup_utils.BackupError as exc:
+            raise exceptions.StorageBackupError(*exc.args) from exc
+
+    def _backup_storage(
```
Comment (danielhollas) [inline src/aiida/storage/sqlite_dos/backend.py:249]: nit: I would move this function before `_backup()` since it has a nice docstring that explains what is going on (helpful when reading from top to bottom).

File: src/aiida/storage/sqlite_dos/backend.py [hunk 5]
```diff
@@ -225,3 +234,36 @@ def _get_mapper_from_entity(entity_type: 'EntityTypes', with_pk: bool):
...
         return mapper, keys
+
+    def _backup(
+        self,
+        dest: str,
+        keep: Optional[int] = None,
+    ):
+        try:
+            backup_manager = backup_utils.BackupManager(dest, keep=keep)
+            backup_manager.backup_auto_folders(lambda path, prev: self._backup_storage(backup_manager, path, prev))
```
Comment (danielhollas) [inline src/aiida/storage/sqlite_dos/backend.py:245]: This line is quite hard to parse. But I am not sure sure how to make it better, it seems the complexity stems from `backup_auto_folders` API.

**Review by danielhollas — APPROVED** [review]
I took this for a spin, since I wanted to try the new `verdi storage backup` command. Works great!

The implementation here seems pretty straightforward, thank you for doing the refactoring that makes it clearer. I only have some nits, my main question is about the profile locking -- is the profile locked when the backup is running?

File: tests/storage/sqlite_dos/test_backend.py (hunk 2 above)
Comment (sphuber) [reply tests/storage/sqlite_dos/test_backend.py:34]: This is because Gio wanted to have support for paths over ssh, so it should support strings of the form `user@some-host:/some/path`. I guess the implementation could also accept `pathlib.Path` values and then convert to string as needed. But the implementation is in `disk-objectstore` and so that would have to be updated first.

File: src/aiida/storage/sqlite_dos/backend.py (hunk 4 above)
Comment (sphuber) [reply src/aiida/storage/sqlite_dos/backend.py:249]: Hmm, I did it this way because `_backup` is the actual abstractmethod that is implemented. The `_backup_storage` is purely an internal method that needed to be separate because of the way `_backup` is implemented. I think I'd rather just add a proper method to `_backup` as well.

File: src/aiida/storage/sqlite_dos/backend.py (hunk 5 above)
Comment (sphuber) [reply src/aiida/storage/sqlite_dos/backend.py:245]: It does. This code is essentially borrowed from the `disk_objectstore` to automatically maintain the folder structure of multiple backup copies where the latest is symlinked and where rsync can make use of incremental backups.

File: tests/storage/sqlite_dos/test_backend.py (hunk 3 above)
Comment (sphuber) [reply tests/storage/sqlite_dos/test_backend.py:37]: Fair.  Will change

File: src/aiida/orm/implementation/storage_backend.py (hunk 1 above)
Comment (danielhollas) [reply src/aiida/orm/implementation/storage_backend.py:416]: Thank you for the explanation. I was essentially concerned if there is apossible  TOCTOU race here.

I guess I still don't understand how are the PID files handled. i.e. where is it ensured that these files are deleted? But that's besides the point here.

File: tests/storage/sqlite_dos/test_backend.py (hunk 2 above)
Comment (danielhollas) [reply tests/storage/sqlite_dos/test_backend.py:34]: Right, to clarify I wanted to suggest that the menthod should accept both so that the logic is not pushed to the callers.

File: src/aiida/orm/implementation/storage_backend.py (hunk 1 above)
Comment (sphuber) [reply src/aiida/orm/implementation/storage_backend.py:416]: To request a lock, it has to be done within a context manager. If the lock is granted, a lock file is written and at the end of the context manager it is automatically cleaned. When a process requests access, it writes a pid file, but since it is not inside a context manager this doesn't do the automatic cleanup. However, when a lock is requested, all existing pid files are checked and are removed if they are stale. So only downside is that pid-files can pile up between lock requests.

File: tests/storage/sqlite_dos/test_backend.py (hunk 2 above)
Comment (sphuber) [reply tests/storage/sqlite_dos/test_backend.py:34]: https://github.com/aiidateam/disk-objectstore/issues/170

**Review by danielhollas — APPROVED** [review]
